File: backend/src/main/java/tom/projects/monkeylog/mapper/ProgramMapper.java
package tom.projects.monkeylog.mapper;

import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;
import org.mapstruct.ReportingPolicy;
import tom.projects.monkeylog.dto.measurement.MeasurementUpdateRequest;
import tom.projects.monkeylog.dto.program.*;
import tom.projects.monkeylog.model.measurement.Measurement;
import tom.projects.monkeylog.model.workout.Program;
import tom.projects.monkeylog.model.workout.ProgramWeek;
import tom.projects.monkeylog.model.workout.Workout;

import java.util.List;

@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE, uses = {WorkoutMapper.class})
public interface ProgramMapper {
    ProgramResponse programToProgramResponse(Program program);
    List<ProgramResponse> programListToProgramResponseList(List<Program> programList);
    Program programCreateRequestToProgram(ProgramCreateRequest programCreateRequest);
    void updateProgram(@MappingTarget Program program, ProgramUpdateRequest programUpdateRequest);
    ProgramWeek programWeekCreateRequestToProgramWeek(ProgramWeekCreateRequest programWeekCreateRequest);
    ProgramWeekResponse programWeekToProgramWeekResponse(ProgramWeek programWeek);
}

File: backendkt/src/main/kotlin/tom/com/monkeylog/mapper/ProgramMapper.kt
package tom.com.monkeylog.mapper

import org.mapstruct.Mapper
import org.mapstruct.MappingTarget
import org.mapstruct.ReportingPolicy
import tom.com.monkeylog.dto.program.*
import tom.com.monkeylog.model.workout.Program
import tom.com.monkeylog.model.workout.ProgramWeek

@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE, uses = [WorkoutMapper::class])
interface ProgramMapper {
    fun programToProgramResponse(program: Program): ProgramResponse
    fun programListToProgramResponseList(programList: List<Program>): List<ProgramResponse>
    fun programCreateRequestToProgram(programCreateRequest: ProgramCreateRequest): Program
    fun updateProgram(@MappingTarget program: Program, programUpdateRequest: ProgramUpdateRequest)
    fun programWeekCreateRequestToProgramWeek(programWeekCreateRequest: ProgramWeekCreateRequest): ProgramWeek
    fun programWeekToProgramWeekResponse(programWeek: ProgramWeek): ProgramWeekResponse
}
